### PR TITLE
MongoDB/PyMongo: Amalgamate PyMongo driver to use CrateDB as backend

### PR DESCRIPTION
**Pull request by amotl**
## About
This patch amalgamates the PyMongo driver to talk to CrateDB instead of a MongoDB server.

## Documentation
Preview: https://cratedb-toolkit--83.org.readthedocs.build/adapter/pymongo.html

## Synopsis
```python
import pymongo
from cratedb_toolkit.adapter.pymongo import PyMongoCrateDBAdapter

with PyMongoCrateDBAdapter(dburi="crate://crate@localhost:4200"):
    client = pymongo.MongoClient("localhost", 27017)
    collection = client.foo.bar

    inserted_id = collection.insert_one({"author": "Mike", "text": "My first blog post!"}).inserted_id
    print(inserted_id)

    document = collection.find({"author": "Mike"})
    print(document)
```

## Status

It can run 95% of the MongoDB "getting started" tutorial successfully.

-- https://pymongo.readthedocs.io/en/stable/tutorial.html
-- [test_pymongo.py](https://github.com/crate-workbench/cratedb-toolkit/blob/d38816df0ba797dfcace43fe57aa4c4928d3fc38/tests/adapter/test_pymongo.py#L199-L294)

## Backlog
- [x] Software tests
- [x] Proper CI setup / configuration
- [x] Documentation

### CODE REVIEW

File: cratedb_toolkit/adapter/pymongo/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+with PyMongoCrateDbAdapter(dburi="crate://crate@localhost:4200"):
+    client = pymongo.MongoClient("localhost", 27017)
+    collection = client.foo.bar
+
+    inserted_id = collection.insert_one({"author": "Mike", "text": "My first blog post!"}).inserted_id
+    print(inserted_id)
+
+    document = collection.find_one({"author": "Mike"})
+    print(document)
+```
```
[the remark this answers is not in the document]
Comment (matriv) [reply cratedb_toolkit/adapter/pymongo/README.md:32]: yep let's go with `DB` both captitals.

[the remark this answers is not in the document]
Comment (amotl) [reply cratedb_toolkit/adapter/pymongo/README.md]: Thanks. Updated with 6d2730a0fa.

File: cratedb_toolkit/adapter/pymongo/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+
+## Examples
+
+To inspect the capabilities of the driver adapter, there is a basic program at
+[pymongo_adapter.py], and test cases at [test_pymongo.py].
```
[the remark this answers is not in the document]
Comment (amotl) [reply cratedb_toolkit/adapter/pymongo/README.md]: Done with 592442ebd5e.

File: cratedb_toolkit/sqlalchemy/patch.py [hunk 3]
```diff
@@ -45,3 +49,11 @@ def default(self, o):
...
                 return o.tolist()
         return json.JSONEncoder.default(self, o)
+
+
+def patch_types_map():
+    """
+    Register missing timestamp data type.
+    """
+    # TODO: Submit patch to `crate-python`.
+    TYPES_MAP["timestamp without time zone"] = sqltypes.TIMESTAMP
```
[the remark this answers is not in the document]
Comment (amotl) [reply cratedb_toolkit/sqlalchemy/patch.py:59]: Added a corresponding note to the backlog.

File: cratedb_toolkit/util/pandas.py
[the remark this answers is not in the document]
Comment (amotl) [reply cratedb_toolkit/util/pandas.py:1]: Added a corresponding note to the backlog.

File: cratedb_toolkit/adapter/pymongo/api.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        # TODO: Upstream to `sqlalchemy-cratedb`.
+        patch_types_map()
+
+        # 2. Prepare pandas.
+        # TODO: Provide unpatching hook.
+        # TODO: Use `with table_kwargs(...)`.
+        from cratedb_toolkit.util.pandas import patch_pandas_sqltable_with_dialect_parameters
+
+        patch_pandas_sqltable_with_dialect_parameters(table_kwargs={"crate_column_policy": "'dynamic'"})
+        patch_pandas_sqltable_with_extended_mapping()
```
[the remark this answers is not in the document]
Comment (amotl) [reply cratedb_toolkit/adapter/pymongo/api.py:63]: Added a corresponding note to the backlog.

File: cratedb_toolkit/adapter/pymongo/collection.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                sql = str(pd_table.table.insert().compile(bind=conn))
+                data = list(data_iter)
+
+                # Invoke amended insert operation, returning the record
+                # identifier as surrogate to MongoDB's `ObjectId`.
+                cursor = conn._dbapi_connection.cursor()
+                cursor.execute(sql=sql + " RETURNING _id", parameters=data[0])
+                outcome = cursor.fetchone()
+                object_id_cratedb = outcome[0]
+                cursor.close()
```
[the remark this answers is not in the document]
Comment (amotl) [reply cratedb_toolkit/adapter/pymongo/collection.py:91]: Added a corresponding note to the backlog.

**Pull request merged**